File: packages/javascript/src/regular-grid.js
const MAX_AXIS_COUNT = 2_147_483_647;
const MAX_SAFE_SIZE = Number.MAX_SAFE_INTEGER;
const MAX_SAFE_SIZE_BIGINT = BigInt(MAX_SAFE_SIZE);
const hasOwn = (value, key) => Object.prototype.hasOwnProperty.call(value, key);

/** Error with one of the stable regular-grid contract codes. */
export class RegularGridError extends Error {
  constructor(code) {
    super(code);
    this.name = "RegularGridError";
    this.code = code;
  }
}

function fail(code) {
  throw new RegularGridError(code);
}

function finiteNumber(value) {
  return typeof value === "number" && Number.isFinite(value);
}

function positiveFiniteNumber(value) {
  return finiteNumber(value) && value > 0;
}

function axisCount(value) {
  return finiteNumber(value) && Number.isSafeInteger(value) && value >= 0 && value <= MAX_AXIS_COUNT;
}

function normalizedZero(value) {
  return value === 0 ? 0 : value;
}

function validateVector(value, predicate) {
  return Array.isArray(value) && value.length === 2 && predicate(value[0]) && predicate(value[1]);
}

function validateInput(params) {
  if (params === null || typeof params !== "object" || Array.isArray(params)) {
    fail("INVALID_INPUT");
  }
  // Inputs are JSON-like records: every own field, including non-enumerable or
  // symbol fields that JSON would otherwise hide, must belong to the four-field shape.
  const keys = Reflect.ownKeys(params);
  if (keys.length !== 4 || !hasOwn(params, "origin") || !hasOwn(params, "spacing") ||
      !hasOwn(params, "columns") || !hasOwn(params, "rows")) {
    fail("INVALID_INPUT");
  }
  if (!validateVector(params.origin, finiteNumber)) {
    fail("INVALID_INPUT");
  }
  if (!validateVector(params.spacing, positiveFiniteNumber)) {
    fail("INVALID_INPUT");
  }
  if (!axisCount(params.columns) || !axisCount(params.rows)) {
    fail("INVALID_INPUT");
  }

  return {
    originX: normalizedZero(params.origin[0]),
    originY: normalizedZero(params.origin[1]),
    spacingX: params.spacing[0],
    spacingY: params.spacing[1],
    columns: normalizedZero(params.columns),
    rows: normalizedZero(params.rows),
  };
}

function coordinate(origin, spacing, axisIndex) {
  const product = axisIndex * spacing;
  const result = origin + product;
  if (!Number.isFinite(product) || !Number.isFinite(result)) {
    fail("COORDINATE_OVERFLOW");
  }
  return normalizedZero(result);
}

function validateIndex(index) {
  if (typeof index !== "number" || !Number.isSafeInteger(index) || index < 0) {
    fail("INVALID_INDEX");
  }
}

function writableNumericArrayAt(out, offset) {
  if (!Array.isArray(out) || out.length < offset + 2) {
    return false;
  }
  function writableSlot(index) {
    const key = String(index);
    const own = Object.getOwnPropertyDescriptor(out, key);
    if (own !== undefined) return "value" in own && own.writable === true;
    if (!Object.isExtensible(out)) return false;
    // Ordinary holes are writable; inherited indexed behavior is outside this route.
    for (let proto = Object.getPrototypeOf(out); proto !== null; proto = Object.getPrototypeOf(proto)) {
      if (Object.getOwnPropertyDescriptor(proto, key) !== undefined) return false;
    }
    return true;
  }
  return writableSlot(offset) && writableSlot(offset + 1);
}

function validateOutput(out, offset) {
  if (typeof offset !== "number" || !Number.isSafeInteger(offset) || offset < 0) {
    fail("INVALID_OUTPUT");
  }
  if (out instanceof Float64Array) {
    if (out.length < offset + 2) {
      fail("INVALID_OUTPUT");
    }
    return;
  }
  if (!writableNumericArrayAt(out, offset)) {
    fail("INVALID_OUTPUT");
  }
}

/**
 * Create an immutable row-major sequence of binary64 planar points.
 * The descriptor stores only origin, spacing, columns, and rows; point storage is caller-owned.
 *
 * Motivating sketches: `2017/Generativos/circlesAlpha` and
 * `2019/generativos/paraisooscuro`. This operation has no approved artistic
 * default or encouraged parameter range: origin, spacing, and point counts are
 * explicit caller choices.
 */
export function regularGrid(params) {
  const values = validateInput(params);
  const exactSize = BigInt(values.columns) * BigInt(values.rows);
  if (exactSize > MAX_SAFE_SIZE_BIGINT) {
    fail("GRID_SIZE_OVERFLOW");
  }
  const size = Number(exactSize);

  if (size !== 0) {
    coordinate(values.originX, values.spacingX, values.columns - 1);
    coordinate(values.originY, values.spacingY, values.rows - 1);
  }

  function coordinatesAt(index) {
    validateIndex(index);
    if (index >= size) {
      fail("INDEX_OUT_OF_RANGE");
    }
    const column = index % values.columns;
    const row = (index - column) / values.columns;
    return [
      coordinate(values.originX, values.spacingX, column),
      coordinate(values.originY, values.spacingY, row),
    ];
  }

  function pointAt(index) {
    return coordinatesAt(index);
  }

  function pointInto(index, out, offset = 0) {
    validateIndex(index);
    if (index >= size) {
      fail("INDEX_OUT_OF_RANGE");
    }
    validateOutput(out, offset);
    const column = index % values.columns;
    const row = (index - column) / values.columns;
    const x = coordinate(values.originX, values.spacingX, column);
    const y = coordinate(values.originY, values.spacingY, row);
    out[offset] = x;
    out[offset + 1] = y;
    return out;
  }

  function serialize() {
    return {
      origin: [values.originX, values.originY],
      spacing: [values.spacingX, values.spacingY],
      columns: values.columns,
      rows: values.rows,
    };
  }

  const sequence = { size, pointAt, pointInto, serialize };
  Object.defineProperty(sequence, "toJSON", { value: serialize });
  return Object.freeze(sequence);
}

File: packages/java/src/main/java/org/procedurals/layout/RegularGrid.java
package org.procedurals.layout;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Compact row-major planar positions; no rendering or random state.
 * Motivated by 2017/Generativos/circlesAlpha (source cell counts require +1)
 * and the extracted lattice in 2019/generativos/paraisooscuro.
 * Origin and positive spacing use caller coordinate units. Counts count points.
 * No artistic default or encouraged range has been established by the survey.
 * Behavior: catalog/operations/regular-grid.json, layout.regular-grid 0.1.0.
 */
public strictfp final class RegularGrid {
    /** Largest supported point count; this is a representation ceiling, not an artistic range. */
    public static final long MAX_SIZE = 9007199254740991L;

    /** Validation or indexed-access failure with a stable catalog error code. */
    public static final class GridException extends IllegalArgumentException {
        /** Stable code: {@code INVALID_INPUT}, {@code GRID_SIZE_OVERFLOW}, {@code COORDINATE_OVERFLOW}, {@code INVALID_INDEX}, {@code INDEX_OUT_OF_RANGE}, or {@code INVALID_OUTPUT}. */
        public final String code;
        /**
         * Creates a stable grid failure.
         *
         * @param code catalog error code
         */
        public GridException(String code) { super(code); this.code = code; }
    }

    private final double ox, oy, dx, dy;
    private final int columns, rows;
    private final long size;

    private RegularGrid(double ox, double oy, double dx, double dy, int columns, int rows, long size) {
        this.ox = ox; this.oy = oy; this.dx = dx; this.dy = dy;
        this.columns = columns; this.rows = rows; this.size = size;
    }

    /**
     * Creates an immutable row-major point sequence from explicit caller-coordinate units.
     * {@code origin} and {@code spacing} are two-number {@code [x,y]} lists; spacing must be
     * positive. {@code columns} and {@code rows} count points, rather than cells, and may be
     * zero. Input values are copied and zero is canonicalized; no artistic default or
     * recommended range is supplied. Motivated by {@code circlesAlpha} and {@code paraisooscuro}
     * as recorded in {@code catalog/operations/regular-grid.json}.
     *
     * @param input exactly {@code origin}, {@code spacing}, {@code columns}, and {@code rows}
     * @return immutable detached grid descriptor without output-sized position storage
     * @throws GridException {@code INVALID_INPUT}, {@code GRID_SIZE_OVERFLOW}, or
     *         {@code COORDINATE_OVERFLOW} when validation fails in contract order
     */
    public static RegularGrid create(Map<String, ?> input) {
        if (input == null || input.size() != 4 || !input.keySet().containsAll(
                Arrays.asList("origin", "spacing", "columns", "rows"))) fail("INVALID_INPUT");
        List<?> origin = vector(input.get("origin"));
        List<?> spacing = vector(input.get("spacing"));
        double ox = number(origin.get(0)), oy = number(origin.get(1));
        double dx = positive(spacing.get(0)), dy = positive(spacing.get(1));
        int columns = count(input.get("columns")), rows = count(input.get("rows"));
        // Products of two nonnegative int32 values fit int64 exactly.
        long size = (long) columns * (long) rows;
        if (size > MAX_SIZE) fail("GRID_SIZE_OVERFLOW");
        if (size != 0) {
            coordinate(ox, dx, columns - 1);
            coordinate(oy, dy, rows - 1);
        }
        return new RegularGrid(ox, oy, dx, dy, columns, rows, size);
    }

    /**
     * Returns the number of points, equal to {@code columns * rows}; it is not a cell count.
     *
     * @return row-major point count
     */
    public long size() { return size; }

    /**
     * Returns detached canonical construction values in catalog-key order. Mutating this map or
     * either nested list cannot change the grid; pass it to {@link #create(Map)} to reconstruct.
     *
     * @return detached {@code origin}, {@code spacing}, {@code columns}, and {@code rows} map
     */
    public Map<String, Object> toMap() {
        Map<String, Object> result = new LinkedHashMap<String, Object>();
        result.put("origin", new ArrayList<Double>(Arrays.asList(ox, oy)));
        result.put("spacing", new ArrayList<Double>(Arrays.asList(dx, dy)));
        result.put("columns", columns); result.put("rows", rows);
        return result;
    }

    /**
     * Returns a fresh binary64 {@code [x,y]} point at a row-major index. X changes fastest;
     * coordinates use caller units and separately rounded multiply then add operations.
     *
     * @param index finite nonnegative safe-integer point index
     * @return detached two-coordinate point
     * @throws GridException {@code INVALID_INDEX} before {@code INDEX_OUT_OF_RANGE}
     */
    public double[] pointAt(long index) {
        validateIndex(index);
        double[] point = new double[2];
        write(index, point, 0);
        return point;
    }

    /**
     * Java-interchange overload for {@link #pointAt(long)} accepting supported numeric carriers.
     *
     * @param index Byte, Short, Integer, Long, Float, or Double safe-integer point index
     * @return detached two-coordinate point
     * @throws GridException {@code INVALID_INDEX} before {@code INDEX_OUT_OF_RANGE}
     */
    public double[] pointAt(Object index) { return pointAt(indexValue(index)); }

    /**
     * Writes exactly two binary64 coordinates after index, range, destination, and offset
     * validation. It retains no destination storage and leaves all destination slots unchanged
     * on failure.
     *
     * @param index finite nonnegative safe-integer point index
     * @param out writable destination requiring two slots from {@code offset}
     * @param offset nonnegative first destination slot
     * @throws GridException {@code INVALID_INDEX}, then {@code INDEX_OUT_OF_RANGE}, then
     *         {@code INVALID_OUTPUT}
     */
    public void pointInto(long index, double[] out, int offset) {
        validateIndex(index);
        if (out == null || offset < 0 || offset > out.length - 2) fail("INVALID_OUTPUT");
        write(index, out, offset);
    }

    /**
     * Java-interchange overload for {@link #pointInto(long, double[], int)}.
     *
     * @param index Byte, Short, Integer, Long, Float, or Double safe-integer point index
     * @param out writable destination requiring two slots from {@code offset}
     * @param offset nonnegative first destination slot
     * @throws GridException indexed and destination failures in the long-overload order
     */
    public void pointInto(Object index, double[] out, int offset) {
        pointInto(indexValue(index), out, offset);
    }

    private void write(long index, double[] out, int offset) {
        long column = index % columns;
        long row = (index - column) / columns;
        double x = coordinate(ox, dx, column), y = coordinate(oy, dy, row);
        out[offset] = x; out[offset + 1] = y;
    }

    private void validateIndex(long index) {
        if (index < 0 || index > MAX_SIZE) fail("INVALID_INDEX");
        if (index >= size) fail("INDEX_OUT_OF_RANGE");
    }

    private static long indexValue(Object value) {
        if (!supportedNumber(value)) { fail("INVALID_INDEX"); return 0; }
        double n = ((Number) value).doubleValue();
        if (!finite(n) || n < 0 || n > MAX_SIZE || n != Math.floor(n)) fail("INVALID_INDEX");
        return (long) n;
    }

    private static List<?> vector(Object value) {
        if (!(value instanceof List) || ((List<?>) value).size() != 2) fail("INVALID_INPUT");
        return (List<?>) value;
    }

    private static double number(Object value) {
        if (!supportedNumber(value)) { fail("INVALID_INPUT"); return 0; }
        double n = ((Number) value).doubleValue();
        if (!finite(n)) fail("INVALID_INPUT");
        return n == 0 ? 0.0 : n;
    }

    private static double positive(Object value) {
        double n = number(value);
        if (!(n > 0)) fail("INVALID_INPUT");
        return n;
    }

    private static int count(Object value) {
        double n = number(value);
        if (n < 0 || n > Integer.MAX_VALUE || n != Math.floor(n)) fail("INVALID_INPUT");
        return (int) n;
    }

    private static double coordinate(double origin, double spacing, long index) {
        double product = (double) index * spacing;
        if (!finite(product)) { fail("COORDINATE_OVERFLOW"); return 0; }
        double value = origin + product;
        if (!finite(value)) fail("COORDINATE_OVERFLOW");
        return value == 0 ? 0.0 : value;
    }

    // Explicit host representations prevent lossy BigDecimal/custom Number coercion.
    private static boolean supportedNumber(Object value) {
        return value instanceof Byte || value instanceof Short || value instanceof Integer
            || value instanceof Long || value instanceof Float || value instanceof Double;
    }

    private static boolean finite(double value) { return !Double.isNaN(value) && !Double.isInfinite(value); }

    private static void fail(String code) { throw new GridException(code); }
}
